sql: add session variable to set schema_locked by default on new tables

fqazi opened the pull request:
This patch will add a new session variable / cluster setting for automatically setting schema_locked:

1. Adds a new session variable `create_table_with_schema_locked` and cluster setting `sql.defaults.create_table_with_schema_locked`, which will create all new tables with schema_locked by default
2. Updates the schema changer to make sure rollback stages are always non-revertible. Previously, this was implicitly done, and is required for having schema_locked in the entire declarative schema changer test suite.
3. Update the sql/schemachanger package to always create tables wit schema_locked. This requires not only updating expected files, but also executing setup statements with the declarative schema changer and adjusting some rollback points for DML injection

Release note (sql change): The create_table_with_schema_locked session variable can be used to control if newly created tables should have schema_locked set.
Informs: #129694
Fixes: #143891

fqazi committed:
sql: add session variable create_table_with_schema_locked

Previously, schema_locked had to be manually set on objects during
creation. This meant that to gain performance improvements on CDC users
would manually have to toggle this on. To address this, this patch adds
a cluster setting and session variable for creating new tables with
schema_locked set.

Informs: #129694
Release note (sql change): Added a new session variable
create_table_with_schema_locked, which can be used to ensure all tables
created by a session have the storage parameter schema_locked set.

fqazi committed:
sql/schemachanger: ensure rollbacks are always non-revertible

Previously, we would plan rollbacks starting from "PostCommitPhase", and
relied on opgen to label things as non-revertible. This would ensure
that majority of plans would get non-revertible implicitly. This can be
problematic if we automatically set/unset schema_locked and certain
plans will no longer start as non-revertible on rollback. To address
this, this patch explicitly ensures that rollbacks start as
non-revertible.

Informs: #129694

Release note: None

fqazi committed:
sql/schemachanger: adopt schema_locked by default in declarative tests

Previously, there were individual tests for confirming if schema_locked
worked correctly in the declarative schema changer. This patch flips
create_table_with_schema_locked and update tests to handle this.
Specifically it does the following:

1) All tables are created as schema_locked in the declarative schema
   changer tests
2) All setup statements are now run via the declarative schema changer
   so that schema_locked is almost never manually toggled
3) Updates all test outputs to expected schema_locked

Informs: #129694
Release note: None

fqazi committed:
sql/schemachanger: handle trigger dependents properly

Previously, trigger dependents used the NotImplementedForPublicObjects
operation to indicate that no operations were needed. Sadly because
triggers are not descriptors, this could lead to assertions in more
complex plans, when schema_locked is involved. To address this, this
patch modifies this operation to accept a trigger ID to detect if the
trigger is dropped.

Release note: None

rafiss reviewed (changes requested):
nice work!

> Reviewed 32 of 840 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-:-ONVlZ_xDYs5NMpwZm8d:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`-- commits` line 11 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVbY225Sog7yWmFeyp:-ONVbY225Sog7yWmFeyq:baw4mvq):*
this should get a release note that describes the session variable. (let's not mention the cluster setting)
___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_subzones/add_column_subzones.side_effects` line 631 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVghxfAFQU_ZT9NR0b:-ONVghxgD36gUoaYVig9:b-l4g0ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/909c825b6973da43497df786041e5fad40560991/pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_subzones/add_column_subzones.side_effects#L631)):*
> ```
> begin transaction #12
> ## PostCommitNonRevertiblePhase stage 3 of 4 with 5 MutationType ops
> upsert descriptor #104
> ```

just curious, this seems fine, but do you know why the order of side effects changed? is it because of the NonRevertible change?
___
*[`pkg/sql/create_table.go` line 2574 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVcxAv3k9QQi7Q9Zcc:-ONVcxAv3k9QQi7Q9Zcd:b-8a8304) ([raw file](https://github.com/cockroachdb/cockroach/blob/31832a6fbe9f941be4e80c012cd8d186781767aa/pkg/sql/create_table.go#L2574)):*
> ```Go
> 	if !ret.IsView() && !ret.IsSequence() &&
> 		params.p.SessionData().CreateTableWithSchemaLocked &&
> 		params.p.IsActive(params.ctx, clusterversion.V25_2) {
> ```

if we add another check here for `p.SessionData().Internal` to skip this for internal executors, could we avoid having to disable it for backup/restore tests?
___
*[`pkg/sql/schemachanger/sctest/decomp.go` line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVeGW4BTOyjJwoQnLv:-ONVeGW4BTOyjJwoQnLw:b-d0n2qx) ([raw file](https://github.com/cockroachdb/cockroach/blob/664648a995f0711b9eaac0eca5cc9efcc8da1b40/pkg/sql/schemachanger/sctest/decomp.go#L47)):*
> ```Go
> 			// We need to disable the declarative schema changer so that we don't end
> 			// up high-fiving ourselves here.
> 			tdb.Exec(t, `SET CLUSTER SETTING sql.defaults.use_declarative_schema_changer = 'off'`)
> ```

why is this getting removed? (and also, do you know why was this even here before?)
___
*[`pkg/sql/catalog/rewrite/rewrite.go` line 1180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVbw9K38E_asPRZcIL:-ONVbw9K38E_asPRZcIM:bsuf6ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/909c825b6973da43497df786041e5fad40560991/pkg/sql/catalog/rewrite/rewrite.go#L1180)):*
> ```Go
> 		}
> 		// Rewrite any function descriptors in the schema change state.
> 		if err := rewriteSchemaChangerState(fnDesc, descriptorRewrites); err != nil {
> ```

how did this turn up? was there a bug lurking here from before?


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143892)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`-- commits` line 11 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVbY225Sog7yWmFeyp:-ONVm6Jd2uGrAtrJ0j8V:br3r26a):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this should get a release note that describes the session variable. (let's not mention the cluster setting)
</blockquote></details>

if you want to wait for the release ntoe until we toggle the default to true, that could be fine too


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143892)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`-- commits` line 11 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVbY225Sog7yWmFeyp:-ONW0KXR9Bx-1zgp9Nry:bw5tuk):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

if you want to wait for the release ntoe until we toggle the default to true, that could be fine too
</blockquote></details>

Done.

Added:     

Release note (sql change): Added a new session variable create_table_with_schema_locked, which can be used to ensure all tables created by a session have the storage parameter schema_locked set.
___
*[`pkg/sql/create_table.go` line 2574 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVcxAv3k9QQi7Q9Zcc:-ONVyt7bCiUqvmEkKaTl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31832a6fbe9f941be4e80c012cd8d186781767aa/pkg/sql/create_table.go#L2574)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

if we add another check here for `p.SessionData().Internal` to skip this for internal executors, could we avoid having to disable it for backup/restore tests?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/sctest/decomp.go` line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVeGW4BTOyjJwoQnLv:-ONW--4h1mw8DpoC-7Vw:bygtyck) ([raw file](https://github.com/cockroachdb/cockroach/blob/664648a995f0711b9eaac0eca5cc9efcc8da1b40/pkg/sql/schemachanger/sctest/decomp.go#L47)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why is this getting removed? (and also, do you know why was this even here before?)
</blockquote></details>

Its being removed so that the entire setup phase uses the declarative schema changer, previously it would run in the legacy schema changer. This prevents schema_locked errors from getting hit. 

The test doesn't use any testing knobs, so this seems completely unnecessary. I'm guessing we had broken fallbacks or something, which is no longer an issue.
___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_subzones/add_column_subzones.side_effects` line 631 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVghxfAFQU_ZT9NR0b:-ONW3YnX0KXLVathkLco:b-76xc1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/909c825b6973da43497df786041e5fad40560991/pkg/ccl/schemachangerccl/testdata/end_to_end/add_column_subzones/add_column_subzones.side_effects#L631)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

just curious, this seems fine, but do you know why the order of side effects changed? is it because of the NonRevertible change?
</blockquote></details>

This is linked to schema_locked. Stage 3 is doing the same but not unsetting job state since its not the final stage. Stage 4 clears all the job stuff and flips schema_locked back on again
___
*[`pkg/sql/catalog/rewrite/rewrite.go` line 1180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVbw9K38E_asPRZcIL:-ONW66HsAuvT4U__-Vh9:b8stnik) ([raw file](https://github.com/cockroachdb/cockroach/blob/909c825b6973da43497df786041e5fad40560991/pkg/sql/catalog/rewrite/rewrite.go#L1180)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

how did this turn up? was there a bug lurking here from before?
</blockquote></details>

Done.

I can't seem to hit this anymore and I wonder if it was just because of a bad testing know when I was developing. Unless we use transactions we really shouldn't need to persist anything in the function descriptor yet. So removing this for now.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143892)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/catalog/rewrite/rewrite.go` line 1180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONVbw9K38E_asPRZcIL:-ONWR9b5-PjPeCT4Xi6i:b-nv5bay) ([raw file](https://github.com/cockroachdb/cockroach/blob/909c825b6973da43497df786041e5fad40560991/pkg/sql/catalog/rewrite/rewrite.go#L1180)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.

I can't seem to hit this anymore and I wonder if it was just because of a bad testing know when I was developing. Unless we use transactions we really shouldn't need to persist anything in the function descriptor yet. So removing this for now.
</blockquote></details>

*bad testing knob


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> Reviewed 1 of 856 files at r5, 1 of 852 files at r9, 1 of 1 files at r10, 10 of 11 files at r11, 10 of 843 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143892)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/schemachanger/sctest/backup.go` line 42 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONowWer2eCDrClTJ3vF:-ONowWer2eCDrClTJ3vG:b-934nnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dabb49ed1c11a574f4fc7fe1ede2e9c12af751f2/pkg/sql/schemachanger/sctest/backup.go#L42)):*
> ```Go
> 	// Disable schema_locked in backup and restore tests, since the userfiles table
> 	// cannot be created with schema_locked by default yet.
> 	factory = factory.WithSchemaLockDisabled()
> ```

now that we're not setting `schema_locked` automatically in the internal executor, can we remove this logic to disable it in backup/restore?


<!-- Sent from Reviewable.io -->

fqazi committed:
catalog/rewrite: rewrite declarative schema changer state for functions

The declarative schema changer state rewrite logic previously skipped
function descriptors. This was an oversight stemming from the fact that
function creation was limited to a single stage. However, with
schema_locked support, function creation involving table references can
now span multiple stages. This patch corrects the oversight by enabling
schema changer state rewriting for functions.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143892)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/sctest/backup.go` line 42 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/143892#-ONowWer2eCDrClTJ3vF:-ONpT8AS0ebpPEKecN5F:b9urheg) ([raw file](https://github.com/cockroachdb/cockroach/blob/dabb49ed1c11a574f4fc7fe1ede2e9c12af751f2/pkg/sql/schemachanger/sctest/backup.go#L42)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

now that we're not setting `schema_locked` automatically in the internal executor, can we remove this logic to disable it in backup/restore?
</blockquote></details>

Done.

Re-enabled all backup and restore tests minus mixed version, where we don't have schema_locked support. 

Also had to revive the function descriptor rewrite change for the `create_function_in_txn` test. This test ran into issues because the plan has two non-revertible stages now with schema_locked.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> Reviewed 1 of 8 files at r1, 1 of 857 files at r16, 2 of 11 files at r18, 1 of 1 files at r19.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143892)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss TFTR!

bors r+
